lib/active_shipping/carriers/correios.rb

```diff
@@ -0,0 +1,276 @@
+          'nVlComprimento' => params[:length],
+          'nVlAltura' => params[:height],
+          'nVlLargura' => params[:width],
+          'nVlDiametro' => params[:diameter],
+          'sCdMaoPropria' => params[:mao_propria_extra],
+          'nVlValorDeclarado' => params[:declared_value_extra],
+          'sCdAvisoRecebimento' => params[:delivery_notice_extra],
+          'nIndicaCalculo' => params[:return_information],
+          'StrRetorno' => params[:return_type]
+        }
+      end
+
+      def query_string(package)
+        params = build_params(package)
+        hash_query_string(params).map { |key, value| "#{CGI.escape(key)}=#{CGI.escape(value.to_s)}" }.join('&')
+      end
+
+      def create_url(package)
+        "#{URL}?#{query_string(package)}"
+      end
+
+      def service_type
+        @options[:services].nil? ? DEFAULT_SERVICES.join(',') : @options[:services].join(',')
+      end
+
+    end
+
+    class CorreiosRateResponse < ActiveShipping::RateResponse
+      attr_reader :raw_responses
+
+      def initialize(success, message, params = {}, options = {})
+        @raw_responses = options[:raw_responses]
+        super
+      end
+    end
+
+    class CorreiosResponse
+
+      def initialize(request, raw_xmls)
+        @request = request
+        @raw_xmls = raw_xmls
+        @documents = raw_xmls.map { |xml| Nokogiri::XML(xml) }
+      end
+
+      def rate_response
+        @rates = rates
+        CorreiosRateResponse.new(true, nil, params_options, response_options)
+      rescue => error
+        CorreiosRateResponse.new(false, error.message, {}, response_options)
+      end
+
+      private
+
+      def response_options
+        {
+          :rates => @rates,
+          :raw_responses => @raw_xmls
+        }
+      end
+
+      def params_options
+        { :responses => @documents }
+      end
+
+      def normalized_services
+        services = @documents.map { |document| document.root.elements }.flatten
+        services = services.group_by do |service_xml|
+          raise(error_message(service_xml)) if error?(service_xml)
+          service_code(service_xml)
+        end
+      end
+
+      def rates_array
+        services = normalized_services.map do |service_id, elements|
+          total_price = elements.sum { |element| price(element) }
+          { :service_code => service_id, :total_price => total_price, :currency => "BRL" }
+        end
+      end
+
+      def rates
+        rates_array.map { |rate_hash| create_rate_estimate(rate_hash) }
+      end
+
+      def create_rate_estimate(rate_hash)
+        RateEstimate.new(@request.origin, @request.destination, Correios.name, AVAILABLE_SERVICES[rate_hash[:service_code]], rate_hash)
+      end
+
+      def error?(xml_item)
+        text = error(xml_item)
+        !text.empty? && text != "0"
+      end
+
+      def error(xml_item)
+        xml_item.css('Erro').text
+      end
+
+      def error_message(xml_item)
+        xml_item.css('MsgErro').text
+      end
```

Comment:
What is the difference between `error?`, `error`, and error_message`? Maybe the naming can be improved here.